processor/src/main/java/org/mapstruct/ap/internal/util/workarounds/EclipseAsMemberOfWorkaround.java

```diff
@@ -0,0 +1,196 @@
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.internal.util.workarounds;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.Element;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+
+import org.eclipse.jdt.core.compiler.CharOperation;
+import org.eclipse.jdt.internal.compiler.apt.dispatch.BaseProcessingEnvImpl;
+import org.eclipse.jdt.internal.compiler.apt.model.ElementImpl;
+import org.eclipse.jdt.internal.compiler.lookup.MethodBinding;
+import org.eclipse.jdt.internal.compiler.lookup.ReferenceBinding;
+import org.eclipse.jdt.internal.compiler.lookup.TypeBinding;
+
+/**
+ * Contains the workaround for {@link Types#asMemberOf(DeclaredType, Element)} using Eclipse implementation types.
+ * <p>
+ * <strong>This class may only be loaded when running within Eclipse</strong>
+ *
+ * @author Andreas Gudian
+ */
+final class EclipseAsMemberOfWorkaround {
+    private EclipseAsMemberOfWorkaround() {
+    }
+
+    /**
+     * Eclipse-specific implementation of {@link Types#asMemberOf(DeclaredType, Element)}.
+     * <p>
+     * Returns {@code null} if the implementation could not determine the result.
+     */
+    static TypeMirror asMemberOf(ProcessingEnvironment environment, DeclaredType containing,
+                                 Element element) {
+
+        ElementImpl elementImpl = tryCast( element, ElementImpl.class );
+        BaseProcessingEnvImpl env = tryCast( environment, BaseProcessingEnvImpl.class );
+
+        if ( elementImpl == null || env == null ) {
+            return null;
+        }
+
+        ReferenceBinding referenceBinding =
+            (ReferenceBinding) ( (ElementImpl) environment.getTypeUtils().asElement( containing ) )._binding;
+
+        MethodBinding methodBinding = (MethodBinding) elementImpl._binding;
+
+        // matches in super-classes have priority
+        MethodBinding inSuperclassHiearchy = findInSuperclassHierarchy( methodBinding, referenceBinding );
+
+        if ( inSuperclassHiearchy != null ) {
+            return env.getFactory().newTypeMirror( inSuperclassHiearchy );
+        }
+
+        // if nothing was found, traverse the interfaces and collect all candidate methods that match
+        List<MethodBinding> candidatesFromInterfaces = new ArrayList<MethodBinding>();
+
+        collectFromInterfaces(
+            methodBinding,
+            referenceBinding,
+            new HashSet<ReferenceBinding>(),
+            candidatesFromInterfaces );
+
+        // there can be multiple matches for the same method name from adjacent interface hierarchies.
+        Collections.sort( candidatesFromInterfaces, MostSpecificMethodBindingComparator.INSTANCE );
```

Comment:
I don't think that it would be enough. You could still implement two unrelated interfaces that each contain a method 'getItem()' with (possibly generic) types that in the end are comparible. E.g one boils down to getItem():Number, the other to getItem():Long.